diff --git a/dist/hooks/useInterceptLinkClicks.js b/dist/hooks/useInterceptLinkClicks.js
index b4e5fd5326e3d01d3eceaef729c10a6120ad3a5d..640f3a79b0ae573a2b548201b9d79acd54f2992b 100644
--- a/dist/hooks/useInterceptLinkClicks.js
+++ b/dist/hooks/useInterceptLinkClicks.js
@@ -122,7 +122,7 @@ function useInterceptLinkClicks({ guardMapRef, }) {
             if (shouldNavigate) {
                 (0, debug_1.debug)('All guards passed, navigating programmatically');
                 // Navigate programmatically since we prevented the default
-                const router = (_a = window.next) === null || _a === void 0 ? void 0 : _a.router;
+                const router = appRouter;
                 if (router) {
                     if (navigateType === 'replace') {
                         router.replace(href);
@@ -169,6 +169,7 @@ function useInterceptLinkClicks({ guardMapRef, }) {
             (0, debug_1.debug)('Cleaning up link click interceptor');
             document.removeEventListener('click', handleLinkClick, true);
             observer.disconnect();
+            isSetup.current = false;
         };
     }, [guardMapRef]);
 }
diff --git a/dist/hooks/useInterceptPopState.js b/dist/hooks/useInterceptPopState.js
index 9fbf9f7b0027fdde96177e20e4802e5dbc14887f..34b97f163038c433ac058d97b2dd227124136893 100644
--- a/dist/hooks/useInterceptPopState.js
+++ b/dist/hooks/useInterceptPopState.js
@@ -32,9 +32,9 @@ function useInterceptPopState({ guardMapRef, }) {
             // NOTE: Called before Next.js router setup which is useEffect().
             // https://github.com/vercel/next.js/blob/50b9966ba9377fd07a27e3f80aecd131fa346482/packages/next/src/client/components/app-router.tsx#L518
             // NOTE: capture on popstate listener is not working on Chrome.
-            window.addEventListener("popstate", onPopState);
+            window.addEventListener("popstate", onPopState, true);
             return () => {
-                window.removeEventListener("popstate", onPopState);
+                window.removeEventListener("popstate", onPopState, true);
             };
         }
     }, [pagesRouter]);
@@ -42,6 +42,7 @@ function useInterceptPopState({ guardMapRef, }) {
 function createHandlePopState(guardMapRef, writeState) {
     let dispatchedState;
     return (nextState) => {
+        nextState = nextState || {};
         const token = nextState.__next_navigation_guard_token;
         const nextIndex = Number(nextState.__next_navigation_guard_stack_index) || 0;
         if (!token || token !== renderedStateRef.current.token) {
